stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -1271,5 +1276,87 @@ const Metadata *swift::findConformingSuperclass(
   return conformingType;
 }
 
+SWIFT_CC(swift)
+void swift::swift_enumerateConformingTypesFromImage(
+  const void *imageAddress,
+  Metadata const *protocolMetadata,
+  TypeEnumerationFunction body,
+  void *bodyContext,
+  SWIFT_CONTEXT void *context,
+  SWIFT_ERROR_RESULT SwiftError **error) {
+
+  // Find the protocol(s) requested by the caller.
+  auto existentialType = dyn_cast<ExistentialTypeMetadata>(protocolMetadata);
+  if (!existentialType) {
+    auto typeName = swift_getTypeName(protocolMetadata, true);
+    swift::fatalError(
+        0,
+        "_enumerateTypes(fromImageAt:conformingTo:) can only test for "
+        "conformance to a protocol, but %.*s is not a protocol.",
+				(int)typeName.length, typeName.data);
+  }
+  const auto& protocols = existentialType->getProtocols();
+
+#if SWIFT_OBJC_INTEROP
+  // Check for any Objective-C protocols in the input. We do not currently
+  // support searching for Objective-C protocol conformances because they do not
+  // produce conformance records in the compiled binary. In the future, if we
+  // detect any such protocols, we can use objc_copyClassList() to add to the
+  // enumerated set (taking care not to enumerate a class twice if it conforms
+  // to an Objective-C protocol *and* a Swift protocol.)
+  for (const auto& protocol : protocols) {
+    if (protocol.isObjC()) {
+      swift::fatalError(
+          0,
+          "_enumerateTypes(fromImageAt:conformingTo:) does not currently "
+          "support Objective-C protocols such as %s.",
+          protocol.getName());
+    }
+  }
+#endif
+
+  // FIXME: capture image address at registration time so we don't need to look them up here?
+  auto snapshot = Conformances.get().SectionsToScan.snapshot();
+  for (const auto& section : snapshot) {
+
+    SymbolInfo info;
+    if (imageAddress) {
+      const auto sectionAddress = section.address();
+      if (!lookupSymbol(sectionAddress, &info)) {
+        // Couldn't find the image this section came from.
+        continue;
+      }
+    }
+
+    if (!imageAddress || imageAddress == info.baseAddress) {
+      // This section corresponds to the supplied image address (or none was
+      // supplied.)
+
+      for (const auto& record : section) {
+        auto protocolDescriptor = record->getProtocol();
+        auto metadata = record->getCanonicalTypeMetadata();
+        if (!protocolDescriptor || !metadata) {
+          // This record does not contain the info we're interested in.
+          continue;
+        }
+
+        for (const auto& protocol : protocols) {
+          const auto& swiftProtocol = protocol.getSwiftProtocol();
+          if (!swift_conformsToProtocol(metadata, swiftProtocol)) {
+            // This type does not conform to the specified protocol(s), so skip.
+            continue;
+          }
+        }
+
+        bool stop = false;
+        body(metadata, &stop, bodyContext, error);
+        if (stop || *error) {
```

Comment:
Do we know `error` is non-null by this point and safe to de-reference?

Reply:
`error` is not even actually a pointer, it's a register. It can never be `NULL`. https://releases.llvm.org/7.0.1/tools/clang/docs/AttributeReference.html#swift-error-result-clang-swift-error-result-clang-swift-error-result

However it could conceivably be pre-populated with a value that is not `NULL`, so I should zero it before starting the loop.